tests/test_upgrade.py

```diff
@@ -0,0 +1,608 @@
+
+    # check current implementation address
+    assert vault_proxy.implementation() == config.vault_impl_addr
+
+    # check vault version
+    assert vault.version() == 3, "version matches"
+
+    # simulate positive rebase and check withdrawal rate
+    tx = lido_oracle_report(cl_diff=ETH(1_000))
+
+    shares_rate_before, shares_rate_after = _shares_rate_from_event(tx)
+    assert shares_rate_after > shares_rate_before, "Shares rate has not increased"
+
+    # rewards are collectable
+    vault.collect_rewards({"from": liquidations_admin})
+
+    ####################
+    # STAGE 2. Upgrade #
+    ####################
+
+    # remember pre-upgrade state for Stage 3
+    preupgrade_vault_steth_balance = steth_token.balanceOf(vault_proxy)
+    preupgrade_finance_steth_balance = steth_token.balanceOf(LIDO_DAO_FINANCE_MULTISIG)
+
+    # deploy a new vault implementation, start and enact DAO vote
+    new_vault_implementation = deploy_vault_and_pass_dao_vote()
+
+    # check configuration
+    assert config.vault_impl_addr != new_vault_implementation.address
+    assert vault_proxy.implementation() == new_vault_implementation.address
+    assert vault.beth_token() == beth_token.address
+    assert vault.steth_token() == steth_token
+    assert vault.admin() == config.lido_dao_agent_address
+    assert vault.version() == 4
+    assert vault.emergency_admin() == ZERO_ADDRESS
+    assert vault.bridge_connector() == ZERO_ADDRESS
+    assert vault.rewards_liquidator() == ZERO_ADDRESS
+    assert vault.insurance_connector() == ZERO_ADDRESS
+    assert vault.anchor_rewards_distributor() == ZERO_BYTES32
+    assert vault.liquidations_admin() == ZERO_ADDRESS
+    assert vault.no_liquidation_interval() == 0
+    assert vault.restricted_liquidation_interval() == 0
+    assert vault.last_liquidation_time() == 0
+    assert vault.last_liquidation_share_price() == 0
+    assert vault.last_liquidation_shares_burnt() == 0
+
+
+    #######################
+    # STAGE 3. 2022/01/26 #
+    #######################
+
+    assert steth_token.balanceOf(vault_proxy) == preupgrade_vault_steth_balance
+    assert (
+        steth_token.balanceOf(LIDO_DAO_FINANCE_MULTISIG)
+        == preupgrade_finance_steth_balance
+    )
+
+    #################
+    # STAGE 4. Mint #
+    #################
+
+    prev_beth_total_supply = beth_token.totalSupply()
+    prev_steth_total_supply = steth_token.totalSupply()
+
+    with brownie.reverts("Minting is discontinued"):
+        vault.submit(
+            deposit_amount,
+            TERRA_ADDRESS,
+            "0x8bada2e",
+            vault.version(),
+            {"from": stranger, "value": deposit_amount},
+        )
+
+    postupgrade_terra_beth_minted_to_stranger = (
+        beth_token.totalSupply() - prev_beth_total_supply
+    )
+    postupgrade_steth_minted_to_stranger = (
+        steth_token.totalSupply() - prev_steth_total_supply
+    )
+
+    assert postupgrade_terra_beth_minted_to_stranger == 0, "no beth was minted"
```

Comment:
minor: can be skipped since there's no possibility of any state change given that the only tx was reverted